--- index.html.tpl.orig	2026-04-26 09:49:22
+++ index.html.tpl	2026-04-26 09:51:37
@@ -136,6 +136,23 @@
             </div>
             </div>
 
+            <!-- Keystrokes -->
+            <input type="image" alt="Keystrokes" src="/novnc/app/images/keystrokes.svg"
+                id="noVNC_keystrokes_button" class="noVNC_button"
+                title="Key Strokes">
+            <div class="noVNC_vcenter">
+            <div id="noVNC_keystrokes" class="noVNC_panel">
+                <div class="noVNC_heading">
+                    <img alt="" src="/novnc/app/images/keystrokes.svg"> Key Strokes
+                </div>
+                <p class="noVNC_subheading">
+                    Only en-US keyboard layout supported.
+                </p>
+                <textarea id="noVNC_keystrokes_text" rows=5></textarea>
+        		<input id="keystroke_send_button" type="button" value="Send" style="background-color: color-mix(in srgb, var(--novnc-lightgrey), transparent 0%);" />
+            </div>
+            </div>
+
             <!-- Toggle fullscreen -->
             <input type="image" alt="Full screen" src="/novnc/app/images/fullscreen.svg"
                 id="noVNC_fullscreen_button" class="noVNC_button noVNC_hidden"
